### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Ayan1089**
* Added config for CI
* Enabled tests
* Added readme for main

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -5,9 +5,9 @@ name: Build
 
 on:
   push:
-    branches: [ main ]
```
Comment (jkrude) [inline .github/workflows/build.yml:8]: Can we use '**' as described [here](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet) to trigger the actions for all branches?

**Review by jkrude — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 1 above)
Comment (Ayan1089) [reply .github/workflows/build.yml:8]: Why do we need so many branches?

Comment (jkrude) [reply .github/workflows/build.yml:8]: That makes it possible for our feature-branches to be tested too. This enables having tested the changes before we merge them.

File: eventseries/src/main/import2Neo4j/ImportData.py [hunk 2]
```diff
@@ -7,7 +7,9 @@
 # from import2Neo4j.Neo4jConnection import Neo4jConnection
 from rdflib import Graph, Literal, URIRef, XSD
 
-class ImportData:
+
+class ImportData(object):
```
Comment (jkrude) [inline eventseries/src/main/import2Neo4j/ImportData.py:11]: There is no reason to inherit from `object` in python 3, right?
https://stackoverflow.com/a/45062077

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        # Calculate cosine similarity between array1 and array2
+        similarity_matrix = cosine_similarity(tfidf_matrix[:len(array1_strings)], tfidf_matrix[len(array1_strings):])
+
+        # Threshold for partial match
+        threshold = self.best_threshold
+
+        # Find partial matches
+        matches = np.argwhere(similarity_matrix >= threshold)
+
+        ctr = 0
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:129]: Unused variable `ctr`

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        partially_matched_events = []
+        # Print partial matches
+        for match in matches:
+            array1_index = match[0]
+            array2_index = match[1]
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:134]: Unused variable `array2_index`

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                self.best_threshold = threshold
+                self.best_f1_score = f1_score
+            # print("F1-Score: ", f1_score)
+            # print("Number of partial matches: ", len(partially_matched_events))
+            # print()
+        print("#####BEST THRESHOLD#####")
+        print(self.best_threshold)
+
+    def wikidata_match(self, existing_matches):
+        partially_matched_events = []
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:104]: Unused variable `partially_matched_events`

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            if f1_score > self.best_f1_score:
+                self.best_threshold = threshold
+                self.best_f1_score = f1_score
+            # print("F1-Score: ", f1_score)
+            # print("Number of partial matches: ", len(partially_matched_events))
+            # print()
+        print("#####BEST THRESHOLD#####")
+        print(self.best_threshold)
+
+    def wikidata_match(self, existing_matches):
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:103]: Please add type-hints for parameter and return value

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                        score = element
+                        new_row = row
+                        new_col = col
+                if (new_row != -1 and new_col != -1):
+                    matches.append([new_row, new_col])
+            #     print(matches)
+
+            # matches = np.argwhere(similarity_matrix >= threshold)
+
+            ctr = 0
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:59]: Unused variable `ctr`

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+import pandas as pd
+import spacy
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.metrics.pairwise import cosine_similarity
+from sklearn.feature_extraction import text
+
+
+class TfIdfMatch:
+    def __init__(self, matches_df: pd.DataFrame) -> None:
+        self.matches_df = matches_df
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:14]: Please add type-hints for class variables

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        self.matches_df.dropna(inplace=True)
+        self.series_titles = matches_df["event_series"].tolist()
+        self.event_titles = matches_df["event"].tolist()
+        self.best_threshold = -1
+        self.best_f1_score = -1
+
+    def matcher(self):
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:21]: Please add docstring outlaying what the method does

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                score = threshold
+                new_row = -1
+                new_col = -1
+                for col in range(num_cols):
+                    element = similarity_matrix[row][col]
+                    if (element >= score):
+                        score = element
+                        new_row = row
+                        new_col = col
+                if (new_row != -1 and new_col != -1):
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:53]: Unnecessary parenthesis

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        threshold_values = [0.5, 0.6, 0.7, 0.8, 0.9]
+        # threshold_values = [0.9]
+
+        # Find partial matches\
+        for threshold in threshold_values:
+            matches = []
+            true_positives = 0
+            false_positives = 0
+            false_negatives = 0
+            num_rows = len(similarity_matrix)
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:41]: I am not sure if `enumerate` is defined for `similarity_matrix`, but this code can be simplified a bit.
```python
for row_idx, row_values in enumerate(similarity_matrix):
                found_col_idx = next((col for col, element in enumerate(row_values) if element >= threshold), -1)
                if found_col_idx != -1:
                    matches.append((row_idx, found_col_idx))
```

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                    matches.append([new_row, new_col])
+            #     print(matches)
+
+            # matches = np.argwhere(similarity_matrix >= threshold)
+
+            ctr = 0
+            partially_matched_events = []
+
+            # Print partial matches
+            for match in matches:
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:63]: Just deconstruct in the iterator `for (array1_index, array2_index) in matches`

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                array1_index = match[0]
+                array2_index = match[1]
+                #         print("Partial match found:")
+                #         print(f"#####EVENT#####{array1_strings[array1_index]}")
+                #         print(f"######SERIES######{array2_strings[array2_index]}")
+
+                if (
+                        self.matches_df.loc[
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:71]: Please break up the complexity of this line

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+            series = json.load(file)
+            series_titles = [item["title"]["value"] for item in series["results"]["bindings"] if "title" in item]
+
+        vectorizer = TfidfVectorizer(stop_words=list(text.ENGLISH_STOP_WORDS))
+        array1_strings = event_titles
+        array2_strings = series_titles
+        tfidf_matrix = vectorizer.fit_transform(event_titles + series_titles)
+
+        # Calculate cosine similarity between array1 and array2
+        similarity_matrix = cosine_similarity(tfidf_matrix[:len(array1_strings)], tfidf_matrix[len(array1_strings):])
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:121]: This code is a near duplicate

File: eventseries/src/main/matcher/tfidf_matcher.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+        # Threshold for partial match
+        threshold = self.best_threshold
```
Comment (jkrude) [inline eventseries/src/main/matcher/tfidf_matcher.py:124]: How should a user of this method know that `matcher()` has to be called before in order to find the best threshold?

**Review by jkrude — CHANGES_REQUESTED** [review]

File: eventseries/src/main/import2Neo4j/ImportData.py (hunk 2 above)
Comment (Ayan1089) [reply eventseries/src/main/import2Neo4j/ImportData.py:11]: @vstark1327 can you check this part, its related to Neo4j

File: eventseries/src/main/matcher/tfidf_matcher.py (hunk 4 above)
Comment (Ayan1089) [reply eventseries/src/main/matcher/tfidf_matcher.py:134]: Let this be, we might need it later for event_series

**Pull request merged**